### PR TITLE
add functions for region selection

### PR DESCRIPTION
**Pull request by nif33**
adds two functions:
-function for fetching regions list from API
-function for adding a new option to a selection list in HTML

these functions are helper functions, they will be tested within the outer function that calls them

### CODE REVIEW

File: src/main/webapp/scripts/region-selector.js [hunk 1]
```diff
@@ -0,0 +1,13 @@
+export { getRegions }
```
Comment (chenzizhao) [inline src/main/webapp/scripts/region-selector.js:1]: nit: Juan: export is usually placed at the end of the file

File: src/main/webapp/scripts/region-selector.js [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+function getRegions() {
+  return gapi.client.youtube.i18nRegions.list({
+    "part": [
+      "snippet"
+    ],
+    "hl": "es_MX"
+  })
+  .then(function(response) {
+    return response.result;
+  });
```
Comment (chenzizhao) [inline src/main/webapp/scripts/region-selector.js:12]: Do you catch the error when you call getRegions()?

File: src/main/webapp/scripts/script.js [hunk 3]
```diff
@@ -8,3 +8,11 @@ function updateDom(text, containerId) {
 function getDomValue(domId) {
   return document.getElementById(domId).value;
 }
+
+function makeSelectList(value, text, selectId) {
```
Comment (chenzizhao) [inline src/main/webapp/scripts/script.js:12]: The function name "makeSelectList" seems to me that it returns an entire select element with all the options set up, while what it actually does is append one option to the select element. What do you think?

File: src/main/webapp/scripts/region-selector.js [hunk 4]
```diff
@@ -0,0 +1,17 @@
+function getRegions() {
+  return gapi.client.youtube.i18nRegions.list({
+    "part": [
+      "snippet"
+    ],
+    "hl": "es_MX"
```
Comment (chenzizhao) [inline src/main/webapp/scripts/region-selector.js:6]: nit: single quote lol

File: src/main/webapp/scripts/region-selector.js [hunk 5]
```diff
@@ -0,0 +1,17 @@
+function getRegions() {
+  return gapi.client.youtube.i18nRegions.list({
+    'part': [
+      'snippet'
+    ],
+    'hl': 'es_MX'
```
Comment (chenzizhao) [inline src/main/webapp/scripts/region-selector.js:6]: nit: maybe add a comment to explain what this field is about?

Comment (juanalonsocruz) [reply src/main/webapp/scripts/region-selector.js:6]: Agreed. Can't tell what this is either.

File: src/main/webapp/scripts/region-selector.js [hunk 6]
```diff
@@ -0,0 +1,17 @@
+function getRegions() {
```
Comment (juanalonsocruz) [inline src/main/webapp/scripts/region-selector.js:1]: You are currently adding "dead code" (code that won't run). When adding functionality, you should also add its usage (unless you creating an API method).  In this case, it is showing the list of regions in the UI. Please add this. Also add test for that.

File: src/main/webapp/scripts/region-selector.js [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+      'snippet'
+    ],
+    'hl': 'es_MX'
+  })
+  .then((response) => {
+    return response.result;
+  })
+  .catch((error) => {
+    console.error('An error occurred getting the region list:', error);
+    throw error;
```
Comment (juanalonsocruz) [inline src/main/webapp/scripts/region-selector.js:13]: doing a console.error and throwing the error will make it have two logs in the console. If the error is informative enough then you don't need to catch. For future PR, maybe you want to catch it to have a different behavior on the UI? Like not letting them submit anything.

File: src/main/webapp/scripts/region-selector.js (hunk 5 above)
Comment (nif33) [reply src/main/webapp/scripts/region-selector.js:6]: I realized I didn't even need it 😅 I don't remember how it got there, probably from some documentation somewhere...

File: src/main/webapp/tests.html [hunk 8]
```diff
@@ -7,16 +7,21 @@
...
   <script src="https://sinonjs.org/releases/sinon-9.0.0.js"></script>
+  <script src="https://apis.google.com/js/api.js"></script>
   <script src="scripts/script.js"></script>
   <script src="scripts/perspective.js"></script>
   <script src="scripts/perspective-api-service.js"></script>
   <script src="scripts/classes/comment-class.js"></script>
   <script src="scripts/tests/perspective-test.js"></script>
-  <script src="https://apis.google.com/js/api.js"></script>
+  <script type="module" src="scripts/tests/popular-test.js"></script>
   <script defer type="module" src="scripts/tests/comment-post-test.js"></script>
```
Comment (FuzzyTree) [inline src/main/webapp/tests.html:25]: is `defer` necessary?

File: src/main/webapp/scripts/script.js [hunk 9]
```diff
@@ -1,4 +1,24 @@
 const API_KEY = '';
+const YOUTUBE_DISCOVERY = 'https://www.googleapis.com/discovery/v1/apis/youtube/v3/rest';
+
+function loadApi() {
+  return new Promise((resolve, reject) => {
+    gapi.load('client', function() {
```
Comment (FuzzyTree) [inline src/main/webapp/scripts/script.js:6]: Nit: please convert all anonymous (unnamed) functions to arrow functions for consistency.

`gapi.load('client', () => {`

File: src/main/webapp/scripts/tests/popular-test.js [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+  items: [
+    { snippet: { gl: 'EG', name: 'Example'}}
+  ]
+}
+
+QUnit.test('region menu display',  function(assert) {
+  const done = assert.async();
+  const getRegions = sinon.fake.returns(
+    Promise.resolve(mockGetRegionsResponse)
+  );
```
Comment (FuzzyTree) [inline src/main/webapp/scripts/tests/popular-test.js:13]: This will not work because `getRegions` is in a different scope from this function. One way to mock getRegions is to import its reference from popular.js:

```
import * as popular from '../popular.js';
.
.
.
let { addRegionOptions, getRegions } = popular;
getRegions = sinon.fake ...
```

You'll also need to export `getRegions` from popular.js

File: src/main/webapp/tests.html [hunk 11]
```diff
@@ -7,16 +7,21 @@
   <link rel="stylesheet" href="https://code.jquery.com/qunit/qunit-2.10.0.css">
 </head>
 <body>
+    <select id="region-select"></select>
   <div id="qunit"></div>
-  <div id="qunit-fixture"></div>
+  <div id="qunit-fixture">
+
+  </div>
```
Comment (FuzzyTree) [inline src/main/webapp/tests.html:14]: Is this change intentional?

Comment (nif33) [reply src/main/webapp/tests.html:14]: going to put the select dropdown menu here once it's working (it's not here right now because this section hides the element and I want to see the box when debugging)

File: src/main/webapp/tests.html (hunk 8 above)
Comment (nif33) [reply src/main/webapp/tests.html:25]: I don't think I need it for mine, @chenzizhao why was it added comment-post-test.js?

Comment (chenzizhao) [reply src/main/webapp/tests.html:25]: ooh sorry it works without defer here my bad, please remove it

File: src/main/webapp/scripts/popular.js [hunk 12]
```diff
@@ -0,0 +1,12 @@
+import { getRegions } from './region-selector.js';
+
+function addRegionOptions() {
```
[the remark this answers is not in the document]
Comment (nif33) [reply src/main/webapp/scripts/popular.js:8]: I don't have access to that link but assuming it is the same info as the public style guide I have added some JSDoc

File: src/main/webapp/scripts/tests/popular-test.js (hunk 10 above)
Comment (nif33) [reply src/main/webapp/scripts/tests/popular-test.js]: It's still not working :/ did I interpret wrong? I am still getting no error message but it is using the real function instead of the fake. I also tried exporting getRegions in popular-test and `import { getRegions } from './tests/popular-test'` in the popular.js file but that caused an error because the functions have the same name. How can I specify when I want to use the getRegions in the original file vs the getRegions fake from the test file...

Comment (FuzzyTree) [reply src/main/webapp/scripts/tests/popular-test.js:13]: Sorry my instructions were not right. Your export in popular.js should be a default export:

`export default {addRegionOptions, getRegions}`

and the import in this file should be 

`import popular from '../popular.js';`

you should use sinon.replace to mock getRegions

```
QUnit.test('region menu display',  function(assert) {
  const done = assert.async();
  const getRegions = sinon.fake.returns(
    Promise.resolve(mockGetRegionsResponse)
  );
  sinon.replace(popular, 'getRegions', getRegions);
  popular.addRegionOptions().then(() => {
    assert.dom('option').hasText('Example');
    done();
  });
});
```

Comment (FuzzyTree) [reply src/main/webapp/scripts/tests/popular-test.js:13]: Sorry, my instructions were not right. This works on my machine:
```
//popular.js
import { getRegions } from './region-selector.js';
let popular;

function addRegionOptions() {
  return loadApi().then(() => popular.getRegions()) // call popular.getRegions() here
  .then((response) => {
    for(const region of response.items) {
      addOptionToSelectList(region.snippet.gl, region.snippet.name, 'region-select');
    }
  });
}   
  
export default popular = {addRegionOptions, getRegions} 
```

```import popular from '../popular.js';

const mockGetRegionsResponse = {
  items: [
    { snippet: { gl: 'EG', name: 'Example'}}
  ]
}   
      
QUnit.test('region menu display',  function(assert) {
  const done = assert.async();
  const getRegions = sinon.fake.returns(
    Promise.resolve(mockGetRegionsResponse)
  );
  sinon.replace(popular, 'getRegions', getRegions);
  popular.addRegionOptions().then(() => {
    assert.dom('option').hasText('Example');
    done();
  });
});```

Comment (FuzzyTree) [reply src/main/webapp/scripts/tests/popular-test.js:13]: `let popular` is necessary because ES6 imports are readonly so you have to use `let` to be able to overwrite the import.

Normally ES6 code is transpiled to ES5 (to support older browsers) where the original instructions should work.

Comment (chenzizhao) [reply src/main/webapp/scripts/tests/popular-test.js:13]: Sorry to cut in -- what's the rationale behind exporting the dependency `export default popular = {addRegionOptions, getRegions}` in popular.js, rather than importing `getRegions` from `./region-selector.js` in popular-test.js?

Comment (FuzzyTree) [reply src/main/webapp/scripts/tests/popular-test.js:13]: Exporting the `getRegions` dependency from popular.js is needed order to get the same reference to `getRegions` that is being used in `popular.js`.

Note that in `popular.js` we're calling `loadApi().then(popular.getRegions)` instead of `loadApi().then(getRegions)` to ensure that the same reference as the export is used.

See the stackoverflow link above for more details.

Comment (chenzizhao) [reply src/main/webapp/scripts/tests/popular-test.js:13]: Thank you! I wasn't clear on the reference of function previously but now I get it.
